### PR TITLE
Only reject routes with neighbor's router-id for ibgp

### PR DESCRIPTION
**Pull request by arifogel**
@progwriter Please add tests. (Not ready for review).

### CODE REVIEW

**Comment by dhalperi** [comment]
Re-add me as reviewer when ready for review.

File: projects/batfish/src/test/resources/org/batfish/grammar/cisco/testrigs/ibgp-only-reject-routerid-match/configs/r1 [hunk 1]
```diff
@@ -0,0 +1,23 @@
+! boot system flash:EOS.swi
+hostname r1
+interface Ethernet1
+   no switchport
+   ip address 10.12.0.1/24
+!
+interface Loopback0
+   ip address 1.0.0.1/32
+!
+interface Loopback1
```
Comment (arifogel) [inline projects/batfish/src/test/resources/org/batfish/grammar/cisco/testrigs/ibgp-only-reject-routerid-match/configs/r1]: Could this interface be superfluous? Is it used any differently than Loopback0?

File: projects/batfish/src/test/resources/org/batfish/grammar/cisco/testrigs/ibgp-only-reject-routerid-match/configs/r1 [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+!
+ip routing
+!
+router bgp 1
+!!! This router ID is the same as r2a,
+!!! but since this is ebgp, r2a should accept the route nevertheless
+   router-id 2.0.0.1
+   neighbor 10.12.0.2 remote-as 2
+   network 1.0.0.1/32
+   network 1.0.0.2/32
```
Comment (arifogel) [inline projects/batfish/src/test/resources/org/batfish/grammar/cisco/testrigs/ibgp-only-reject-routerid-match/configs/r1]: See earlier comment about Loopback1.

File: projects/batfish/src/test/java/org/batfish/bdp/BdpDataPlanePluginTest.java [hunk 3]
```diff
@@ -177,4 +177,30 @@ public void testIbgpRejectOwnAs() throws IOException {
...
+    Set<Prefix> r2aPrefixes =
+        r2aRoutes.stream().map(r -> r.getNetwork()).collect(Collectors.toSet());
+    Set<Prefix> r2bPrefixes =
+        r2bRoutes.stream().map(r -> r.getNetwork()).collect(Collectors.toSet());
+    Prefix r1Loopback0Prefix = new Prefix("1.0.0.1/32");
+    Prefix r1Loopback1Prefix = new Prefix("1.0.0.2/32");
+    assertTrue(r2aPrefixes.contains(r1Loopback0Prefix));
+    assertTrue(r2aPrefixes.contains(r1Loopback1Prefix));
+    assertTrue(r2bPrefixes.contains(r1Loopback0Prefix));
+    assertTrue(r2bPrefixes.contains(r1Loopback1Prefix));
```
Comment (arifogel) [inline projects/batfish/src/test/java/org/batfish/bdp/BdpDataPlanePluginTest.java]: What do we learn from the presence of the advertised prefixes on `r2b`?

File: projects/batfish/src/test/java/org/batfish/bdp/BdpDataPlanePluginTest.java [hunk 4]
```diff
@@ -177,4 +177,59 @@ public void testIbgpRejectOwnAs() throws IOException {
      */
     assertFalse(r2bPrefixes.contains(r1Loopback0Prefix));
   }
+
+  @Test
+  public void testEbgpAcceptSameNeighborID() throws IOException {
+    String testrigName = "ibgp-only-reject-routerid-match";
```
Comment (arifogel) [inline projects/batfish/src/test/java/org/batfish/bdp/BdpDataPlanePluginTest.java]: The name of this testrig is not meaningfully different from the name of the other testrig. Perhaps name them after the test function names?

File: projects/batfish/src/test/resources/org/batfish/grammar/cisco/testrigs/ibgp-only-reject-routerid-match/configs/r1 [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+!
+interface Loopback0
+   ip address 1.0.0.1/32
+!
+ip routing
+!
+router bgp 1
+!!! This router ID is the same as r3
+   router-id 9.9.9.9
+   neighbor 10.12.0.2 remote-as 2
```
Comment (arifogel) [inline projects/batfish/src/test/resources/org/batfish/grammar/cisco/testrigs/ibgp-only-reject-routerid-match/configs/r1]: Please peer on loopbacks for iBGP. Use static routes for each loopback that must be reached as a neighbor address.

File: projects/batfish/src/test/resources/org/batfish/grammar/cisco/testrigs/ibgp-only-reject-routerid-match/configs/r2 [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+!
+interface Loopback0
+   ip address 2.0.0.1/32
+!
+ip route 2.0.0.2/32 10.22.0.2
+!
+ip routing
+!
+route-map 1-to-2 permit 200
+route-map 3-to-2 permit 200
```
Comment (arifogel) [inline projects/batfish/src/test/resources/org/batfish/grammar/cisco/testrigs/ibgp-only-reject-routerid-match/configs/r2]: These route-maps and their applications in the bgp stanza serve no purpose.

File: projects/batfish/src/test/resources/org/batfish/grammar/cisco/testrigs/ibgp-reject-routerid-match/configs/r1 [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+!
+interface Loopback0
+   ip address 1.0.0.1/32
+!
+ip routing
+!
+router bgp 1
+!!! This router ID is the same as r3
+   router-id 9.9.9.9
+   neighbor 10.14.0.4 remote-as 1
```
Comment (arifogel) [inline projects/batfish/src/test/resources/org/batfish/grammar/cisco/testrigs/ibgp-reject-routerid-match/configs/r1]: Please use loopbacks to establish iBGP sessions. Add static routes to those loopbacks wherever they are referred to. The next hop should be the physical interface address of the neighbor.

File: projects/batfish/src/test/resources/org/batfish/grammar/cisco/testrigs/ibgp-reject-routerid-match/configs/r2 [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+!
+interface Loopback0
+   ip address 2.0.0.1/32
+!
+ip routing
+!
+!
+router bgp 1
+   router-id 2.2.2.2
+   neighbor 10.24.0.4 remote-as 1
```
Comment (arifogel) [inline projects/batfish/src/test/resources/org/batfish/grammar/cisco/testrigs/ibgp-reject-routerid-match/configs/r2]: Please use loopbacks to establish iBGP sessions. Add static routes to those loopbacks wherever they are referred to. The next hop should be the physical interface address of the neighbor.

File: projects/batfish/src/test/resources/org/batfish/grammar/cisco/testrigs/ibgp-reject-routerid-match/configs/r3 [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+!
+interface Loopback0
+   ip address 3.0.0.1/32
+!
+ip routing
+!
+router bgp 1
+!!! same ID as r1
+   router-id 9.9.9.9
+   neighbor 10.34.0.4 remote-as 1
```
Comment (arifogel) [inline projects/batfish/src/test/resources/org/batfish/grammar/cisco/testrigs/ibgp-reject-routerid-match/configs/r3]: Please use loopbacks to establish iBGP sessions. Add static routes to those loopbacks wherever they are referred to. The next hop should be the physical interface address of the neighbor.

File: projects/batfish/src/test/resources/org/batfish/grammar/cisco/testrigs/ibgp-reject-routerid-match/configs/r4 [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+   ip address 10.34.0.4/24
+!
+interface Loopback0
+   ip address 4.0.0.1/32
+!
+ip routing
+!
+router bgp 1
+   router-id 4.4.4.4
+   neighbor 10.14.0.1 remote-as 1
```
Comment (arifogel) [inline projects/batfish/src/test/resources/org/batfish/grammar/cisco/testrigs/ibgp-reject-routerid-match/configs/r4]: Please use loopbacks to establish iBGP sessions. Add static routes to those loopbacks wherever they are referred to. The next hop should be the physical interface address of the neighbor.

File: projects/batfish/src/test/resources/org/batfish/grammar/cisco/testrigs/ebgp-accept-routerid-match/configs/r3 [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+hostname r3
+!
+interface Ethernet1
+   no switchport
+   ip address 10.23.0.3/24
+!
+interface Loopback0
+   ip address 3.0.0.3/32
+!
+ip route 2.0.0.1/32 10.22.0.1
```
Comment (arifogel) [inline projects/batfish/src/test/resources/org/batfish/grammar/cisco/testrigs/ebgp-accept-routerid-match/configs/r3]: This is not necessary for eBGP.

File: projects/batfish/src/test/resources/org/batfish/grammar/cisco/testrigs/ebgp-accept-routerid-match/configs/r2 [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+!
+interface Loopback0
+   ip address 2.0.0.1/32
+!
+ip route 2.0.0.2/32 10.22.0.2
+!
+ip routing
+!
+route-map 1-to-2 permit 200
+route-map 3-to-2 permit 200
```
Comment (arifogel) [inline projects/batfish/src/test/resources/org/batfish/grammar/cisco/testrigs/ebgp-accept-routerid-match/configs/r2]: Please remove these route-maps and their references below.

File: projects/batfish/src/test/resources/org/batfish/grammar/cisco/testrigs/ibgp-reject-routerid-match/configs/r4 [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+   neighbor 2.0.0.1 remote-as 1
+   neighbor 2.0.0.1 update-source Loopback0
+   neighbor 2.0.0.1 route-reflector-client
+   neighbor 3.0.0.1 remote-as 1
+   neighbor 3.0.0.1 update-source Loopback0
+   neighbor 3.0.0.1 route-reflector-client
+!
+!ip route 1.0.0.1/32 10.14.0.1
+!ip route 2.0.0.1/32 10.24.0.2
+!ip route 3.0.0.1/32 10.34.0.3
```
Comment (arifogel) [inline projects/batfish/src/test/resources/org/batfish/grammar/cisco/testrigs/ibgp-reject-routerid-match/configs/r4]: These routes should not be commented.

**Comment by arifogel** [comment]
Looks good. Sending to @dhalperi for review.

File: projects/batfish/src/main/java/org/batfish/bdp/VirtualRouter.java [hunk 14]
```diff
@@ -820,7 +820,8 @@ public int propagateBgpRoutes(Map<String, Node> nodes, Map<Ip, Set<String>> ipOw
 
               Ip remoteOriginatorIp = bgpRemoteRoute.getOriginatorIp();
               // don't accept routes whose originator ip is my BGP id
```
Comment (dhalperi) [inline projects/batfish/src/main/java/org/batfish/bdp/VirtualRouter.java]: update comment? external BGP routes?

File: projects/batfish/src/test/java/org/batfish/bdp/BdpDataPlanePluginTest.java [hunk 15]
```diff
@@ -148,6 +148,30 @@ public void testApplySourceNatInvalidAclThrows() {
...
+    SortedMap<String, SortedMap<String, SortedSet<AbstractRoute>>> routes =
+        dataPlanePlugin.getRoutes();
+    SortedSet<AbstractRoute> r1Routes = routes.get("r1").get(Configuration.DEFAULT_VRF_NAME);
+    SortedSet<AbstractRoute> r3Routes = routes.get("r3").get(Configuration.DEFAULT_VRF_NAME);
+    Set<Prefix> r1Prefixes = r1Routes.stream().map(r -> r.getNetwork()).collect(Collectors.toSet());
+    Set<Prefix> r3Prefixes = r3Routes.stream().map(r -> r.getNetwork()).collect(Collectors.toSet());
+    Prefix r1Loopback0Prefix = new Prefix("1.0.0.1/32");
+    Prefix r3Loopback0Prefix = new Prefix("3.0.0.3/32");
+    // Ensure that r3loopback was accepted by r1
+    assertTrue(r1Prefixes.contains(r3Loopback0Prefix));
```
Comment (dhalperi) [inline projects/batfish/src/test/java/org/batfish/bdp/BdpDataPlanePluginTest.java]:
```java
assertThat(r1Prefixes, contains(r3Loopback0prefix))
```

Will produce better error messages.

(e.g.,

"Expecting an iterable containing X, but was an iterable: [a,b,c]"

File: projects/batfish/src/test/java/org/batfish/bdp/BdpDataPlanePluginTest.java [hunk 16]
```diff
@@ -181,6 +205,30 @@ public void testIbgpRejectOwnAs() throws IOException {
...
+    SortedSet<AbstractRoute> r2Routes = routes.get("r2").get(Configuration.DEFAULT_VRF_NAME);
+    SortedSet<AbstractRoute> r3Routes = routes.get("r3").get(Configuration.DEFAULT_VRF_NAME);
+    Set<Prefix> r2Prefixes = r2Routes.stream().map(r -> r.getNetwork()).collect(Collectors.toSet());
+    Set<Prefix> r3Prefixes = r3Routes.stream().map(r -> r.getNetwork()).collect(Collectors.toSet());
+    // 9.9.9.9/32 is the prefix we test with
+    Prefix r1AdvertisedPrefix = new Prefix("9.9.9.9/32");
+    // Ensure that the prefix is accepted by r2, because router ids are different
+    assertTrue(r2Prefixes.contains(r1AdvertisedPrefix));
+    // Ensure that the prefix is rejected by r3, because router ids are the same
+    assertFalse(r3Prefixes.contains(r1AdvertisedPrefix));
```
Comment (dhalperi) [inline projects/batfish/src/test/java/org/batfish/bdp/BdpDataPlanePluginTest.java]: same assert questions as before

File: projects/batfish/src/test/java/org/batfish/bdp/BdpDataPlanePluginTest.java [hunk 17]
```diff
@@ -181,6 +205,30 @@ public void testIbgpRejectOwnAs() throws IOException {
...
+            TESTRIGS_PREFIX + testrigName, configurationNames, null, null, null, null, _folder);
+    BdpDataPlanePlugin dataPlanePlugin = new BdpDataPlanePlugin();
+    dataPlanePlugin.initialize(batfish);
+    dataPlanePlugin.computeDataPlane(false);
+    SortedMap<String, SortedMap<String, SortedSet<AbstractRoute>>> routes =
+        dataPlanePlugin.getRoutes();
+    SortedSet<AbstractRoute> r2Routes = routes.get("r2").get(Configuration.DEFAULT_VRF_NAME);
+    SortedSet<AbstractRoute> r3Routes = routes.get("r3").get(Configuration.DEFAULT_VRF_NAME);
+    Set<Prefix> r2Prefixes = r2Routes.stream().map(r -> r.getNetwork()).collect(Collectors.toSet());
+    Set<Prefix> r3Prefixes = r3Routes.stream().map(r -> r.getNetwork()).collect(Collectors.toSet());
```
Comment (dhalperi) [inline projects/batfish/src/test/java/org/batfish/bdp/BdpDataPlanePluginTest.java]: Style advice: I like to structure my tests into 2/3 clear sections.

1. Initialization ("nothing interesting here") and constants.
2. The actual code that is under test. As short as possible.
3. Assertions. (perhaps merged with 2, if 2 is one line.)

I usually have whitespace and/or comments between them.

Also, make use of helpers for any common code. (see `makeFlow`, `makeAcl`, etc.)

An example: https://github.com/batfish/batfish/commit/fd022f5d46fae6670758245fa747e5e707d59d85#diff-7ee7600d68323ad23eae06bf4434516cR71

I learned this over time as a pattern that makes tests readable -- all the "background" is at the top and all the "tests" are at the bottom. Whitespace helps.

Feel free to go your own way.

**Pull request merged**